### PR TITLE
Implement `TryFromBytes` for `[T]`

### PR DESCRIPTION
**Pull request by jswrenn**
Makes progress on #5.

### CODE REVIEW

File: src/lib.rs [hunk 1]
```diff
@@ -3524,6 +3524,30 @@ safety_comment! {
...
+        // then by definition it holds for each element of the type. Thus, since
+        // the caller has promised that it holds of the entire `[T]`, it must
+        // also hold for each individual `T`.
+        //
+        // Thus, the preconditions for this call are satisfied.
+        //
+        // Note that this call may panic, but it would still be sound even if it
+        // did. `is_bit_valid` does not promise that it will not panic (in fact,
+        // it explicitly warns that it's a possibility), and we have not
+        // violated any safety invariants that we must fix before returning.
```
Comment (joshlf) [inline src/lib.rs:3548]: Did you mean to update this?

File: src/util.rs [hunk 2]
```diff
@@ -286,6 +290,73 @@ pub(crate) mod ptr {
...
+                //
+                // We satisfy all three of these conditions here:
+                // 1. `base` (by invariant on `self`) points to an allocated
+                //    object. By contract, `self.len()` accurately reflects the
+                //    number of elements in the slice. `i` is in bounds of
+                //   `c.len()` by construction, and so the result of this
+                //   addition cannot overflow past the end of the allocation
+                //   referred to by `c`.
+                // 2. By invariant on `Ptr`, `self` addresses a byte range whose
+                //    lengths fits in an `isize`. Since `elem` is contained by
```
Comment (joshlf) [inline src/util.rs:320]:
```suggestion
                //    length fits in an `isize`. Since `elem` is contained by
```

Comment (joshlf) [inline src/util.rs:320]:
```suggestion
                //    lengths fits in an `isize`. Since `elem` is contained in
```

File: src/util.rs [hunk 3]
```diff
@@ -286,6 +290,73 @@ pub(crate) mod ptr {
...
+                //    `self`, the computed offset of `elem` must fit within
+                //    `isize.`
+                // 3. By invariant on `Ptr`, `self` addresses a byte range which
+                //    does not wrap around the address space. Since `elem` is
+                //    contained by `self`, the computed offset of `elem` must
```
Comment (joshlf) [inline src/util.rs:325]:
```suggestion
                //    contained in `self`, the computed offset of `elem` must
```

File: src/util.rs [hunk 4]
```diff
@@ -286,6 +290,73 @@ pub(crate) mod ptr {
...
+                //    must not overflow or wrap around, so `elem >= base > 0`.
+                let elem = unsafe { NonNull::new_unchecked(elem) };
+
+                // SAFETY: The safety invariants of `Ptr` (see definition) are
+                // satisfied:
+                // 1. `elem` is derived from a valid Rust allocation, because
+                //    `self` is derived from a valid Rust allocation, by
+                //    invariant on `Ptr`
+                // 2. `elem` has the same provenance as `self`, because it is
+                //    within the allocation of `self`
```
Comment (joshlf) [inline src/util.rs:341]: I believe you also need to justify that all of the operations used to derive `elem` from `self` are provenance-preserving.

**Review by joshlf — CHANGES_REQUESTED** [review]

File: src/lib.rs (hunk 1 above)
Comment (jswrenn) [reply src/lib.rs:3548]: Whoops, yep. Forgot this one.

File: src/util.rs (hunk 4 above)
Comment (jswrenn) [reply src/util.rs:341]: Hm, [`pointer::add`](https://doc.rust-lang.org/std/primitive.pointer.html#method.add) doesn't yet document that it preserves provenance. Got another angle of attack for this, or shall we leave it as a TODO and file a PR to Rust?

Comment (joshlf) [reply src/util.rs:341]: TODO seems fine to me, especially since there's discussion of stabilizing strict provenance.

File: src/lib.rs [hunk 5]
```diff
@@ -3524,6 +3524,42 @@ safety_comment! {
...
+        // not panic (in fact, it explicitly warns that it's a possibility), and
+        // we have not violated any safety invariants that we must fix before
+        // returning.
+        c.iter().all(|elem|
+            // SAFETY: We uphold the safety contract of `is_bit_valid(elem)`, by
+            // precondition on the surrounding call to `is_bit_valid`. The
+            // memory referenced by `elem` is contained entirely within `c`, and
+            // satisfies the preconditions satisfied by `c`. By axiom, we assume
+            // that `Iterator:all` does not invalidate these preconditions
+            // (e.g., by writing to `elem`.) Since `elem` is derived from `c`,
```
Comment (joshlf) [inline src/lib.rs:3557]:
```suggestion
            // (e.g., by writing to `elem`). Since `elem` is derived from `c`,
```

**Review by joshlf — APPROVED** [review]

**Pull request merged**